Java frontend: get qualified generic types of static fields

smowton opened the pull request:
Before it would unintentionally erase SomeClass.field from `A<B>` to just the unqualified A type.

Needs tests adding, of course.

smowton committed:
Java frontend: get qualified generic types of static fields

Before it would unintentionally erase SomeClass.field from A<B> to just the unqualified A type.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 81b3148248c1ecf4deaa2f7cbba65f1759f4445b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100277615

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1554,12 +1554,10 @@ code_blockt java_bytecode_convert_methodt::convert_instructions(
       const symbol_exprt symbol_expr(
-        get_static_field(arg0.get_string(ID_class), field_name), arg0.type());
-
-      INVARIANT(
```
:question: This looks like useful context for this symbol not existing that goes away by folding it into `lookup_ref` - do you disagree?

smowton replied:
Seems pointless to have an invariant that can never fail, and I think a symbol table lookup returning a reference is widely recognised as find-or-die?

thk123 replied:
My point is `"getstatic symbol should have been created before method conversion"` is more useful than simply a symbol not existing. Like without that message you might (wrongly) conclude that this conversion should actually be adding to the symbol table

I'm suggesting the invariant goes about the call to `lookup_ref` in case that wasn't clear. Like,

smowton replied:
Added a clarifying comment

thk123 reviewed (commented):
As you say - this of course needs a test

smowton committed:
Add test for generic static field type consistency

smowton commented:
@thk123 test added

smowton committed:
Comment that static field symbols should exist

Avoid the suggestion that `get_static_field` might be expected to create them -- it doesn't,
java_bytecode_convert_class does and we should be able to safely assume they exist
during convert_method.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a208cce75f4a8a72ad7138d879c8e6317c62b051).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100610260

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 35213c49f35691285d2af9771da5c9e2de0cd8ba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100644186
